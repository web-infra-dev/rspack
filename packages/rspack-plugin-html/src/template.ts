/**
 * The following code is modified based on
 * https://github.com/jantimon/html-webpack-plugin/blob/d5ce5a8f2d12a2450a65ec51c285dd54e36cd921/index.js
 * https://github.com/jantimon/html-webpack-plugin/blob/d5ce5a8f2d12a2450a65ec51c285dd54e36cd921/lib/loader.js
 *
 * MIT Licensed
 * Author Jan Nicklas
 * Copyright (c) JS Foundation and other contributors
 * https://github.com/jantimon/html-webpack-plugin/blob/d5ce5a8f2d12a2450a65ec51c285dd54e36cd921/LICENSE
 */
import vm from "vm";

/**
 * compile template to js code.
 */
export interface TemplateCompiler<T = any> {
	compile(
		content: string,
		options?: {
			filename: string;
		} & T
	): Promise<string>;
	options?: T;
}

export const defaultTemplateCompiler: TemplateCompiler = {
	async compile(content, options) {
		const template = (await import("lodash.template")).default(content, {
			interpolate: /<%=([\s\S]+?)%>/g,
			variable: "data",
			...options
		});
		return `function template(templateParams) { with(templateParams) { return (${template.source})(); } }\ntemplate`;
	},
	options: {}
};

/**
 * eval js code to js function or js string.
 */
export async function evaluate(
	compiled: string,
	publicPath: string,
	templateFilename: string
): Promise<string | (() => string | Promise<string>)> {
	if (!compiled) {
		return Promise.reject(
			new Error("The templateCompiler didn't provide a compiled result")
		);
	}
	const vmContext = vm.createContext({
		...global,
		process,
		HTML_WEBPACK_PLUGIN: true,
		require: require,
		htmlWebpackPluginPublicPath: publicPath,
		URL: require("url").URL,
		__filename: templateFilename
	});
	const vmScript = new vm.Script(compiled, { filename: templateFilename });
	// Evaluate code and cast to string
	let newSource;
	try {
		newSource = vmScript.runInContext(vmContext);
	} catch (e) {
		return Promise.reject(e);
	}
	return typeof newSource === "string" || typeof newSource === "function"
		? Promise.resolve(newSource)
		: Promise.reject(
				new Error(
					'The compiled template "' +
						templateFilename +
						"\" didn't return html."
				)
			);
}
