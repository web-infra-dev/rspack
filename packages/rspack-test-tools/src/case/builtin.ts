import {
	IRspackBuiltinProcessorOptions,
	RspackBuiltinProcessor
} from "../processor";
import { BasicCaseCreator } from "../test/creator";
import path from "path";

const FILTERS: Record<
	string,
	IRspackBuiltinProcessorOptions["snapshotFileFilter"]
> = {
	"plugin-css": (file: string) => file.endsWith(".css"),
	"plugin-css-modules": (file: string) =>
		file.endsWith(".css") ||
		(file.endsWith(".js") && !file.includes("runtime")),
	"plugin-html": (file: string) => file.endsWith(".html")
};

const creator = new BasicCaseCreator({
	clean: true,
	describe: false,
	description(name) {
		return `${name} should match snapshot`;
	},
	steps: ({ name, src }) => {
		const cat = path.basename(path.dirname(src));
		const filter = FILTERS[cat];
		return [
			new RspackBuiltinProcessor({
				name,
				snapshot: "output.snap.txt",
				snapshotFileFilter: filter
			})
		];
	}
});

export function createBuiltinCase(name: string, src: string, dist: string) {
	creator.create(name, src, dist);
}
