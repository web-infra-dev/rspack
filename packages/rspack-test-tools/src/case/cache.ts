import { CacheProcessor } from "../processor/cache";
import { CacheRunnerFactory } from "../runner";
import { BasicCaseCreator } from "../test/creator";
import { ECompilerType, type TCompilerOptions } from "../type";

type TTarget = TCompilerOptions<ECompilerType.Rspack>["target"];

const creators: Map<
	TTarget,
	BasicCaseCreator<ECompilerType.Rspack>
> = new Map();

function getCreator(target: TTarget) {
	if (!creators.has(target)) {
		creators.set(
			target,
			new BasicCaseCreator({
				clean: true,
				describe: true,
				target,
				steps: ({ name, target }) => [
					new CacheProcessor({
						name,
						target: target as TTarget,
						compilerType: ECompilerType.Rspack,
						configFiles: ["rspack.config.js", "webpack.config.js"]
					})
				],
				runner: CacheRunnerFactory
			})
		);
	}
	return creators.get(target)!;
}

export function createCacheCase(
	name: string,
	src: string,
	dist: string,
	target: TCompilerOptions<ECompilerType.Rspack>["target"]
) {
	const creator = getCreator(target);
	creator.create(name, src, dist);
}
