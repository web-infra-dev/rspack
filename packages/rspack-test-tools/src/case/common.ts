import fs from "fs-extra";
import path from "path";
import merge from "webpack-merge";
import { readConfigFile } from "../helper";
import { normalizePlaceholder } from "../helper/expect/placeholder";
import checkArrayExpectation from "../helper/legacy/checkArrayExpectation";
import {
	ECompilerType,
	type ITestCompilerManager,
	type ITestContext,
	type ITestEnv,
	type TCompilation,
	type TCompiler,
	type TCompilerOptions,
	type TCompilerStats
} from "../type";

export function getCompiler<T extends ECompilerType = ECompilerType.Rspack>(
	context: ITestContext,
	name: string
) {
	return context.getCompiler(
		name,
		ECompilerType.Rspack
	) as ITestCompilerManager<T>;
}

export async function config<T extends ECompilerType = ECompilerType.Rspack>(
	context: ITestContext,
	name: string,
	configFiles: string[],
	defaultOptions: TCompilerOptions<T> = {}
): Promise<TCompilerOptions<T>> {
	const compiler = getCompiler<T>(context, name);
	compiler.setOptions(defaultOptions);
	if (Array.isArray(configFiles)) {
		const fileOptions = readConfigFile<T>(
			configFiles.map(i => context.getSource(i)),
			context,
			defaultOptions
		)[0];
		compiler.mergeOptions(fileOptions);
	}
	return compiler.getOptions() as TCompilerOptions<T>;
}

export async function compiler<T extends ECompilerType = ECompilerType.Rspack>(
	context: ITestContext,
	name: string
): Promise<TCompiler<T>> {
	const compiler = getCompiler(context, name);
	compiler.createCompiler();
	return compiler.getCompiler()! as TCompiler<T>;
}

export async function build<T extends ECompilerType = ECompilerType.Rspack>(
	context: ITestContext,
	name: string
): Promise<TCompiler<T>> {
	const compiler = getCompiler(context, name);
	await compiler.build();
	return compiler.getCompiler()! as TCompiler<T>;
}

export async function run<T extends ECompilerType = ECompilerType.Rspack>(
	env: ITestEnv,
	context: ITestContext,
	name: string,
	findBundle: (
		context: ITestContext,
		options: TCompilerOptions<T>
	) => string[] | string | void
) {
	const testConfig = context.getTestConfig();
	if (testConfig.noTests) return;

	const compiler = getCompiler(context, name);
	if (typeof testConfig.beforeExecute === "function") {
		testConfig.beforeExecute(compiler.getOptions());
	}

	let bundles: string[] | void | string;
	if (testConfig.bundlePath) {
		bundles = testConfig.bundlePath;
	} else if (typeof findBundle === "function") {
		bundles = findBundle(context, compiler.getOptions() as TCompilerOptions<T>);
	} else {
		bundles = [];
	}

	if (typeof bundles === "string") {
		bundles = [bundles];
	}
	if (!bundles || !bundles.length) {
		return;
	}

	for (const bundle of bundles!) {
		if (!bundle) {
			continue;
		}
		const runner = context.getRunner(name, bundle, env);
		const mod = runner.run(bundle);
		const result =
			context.getValue<Array<Promise<unknown>>>(name, "modules") || [];
		result.push(mod);
		context.setValue<Array<Promise<unknown>>>(name, "modules", result);
	}

	const results =
		context.getValue<Array<Promise<unknown>>>(name, "modules") || [];
	await Promise.all(results);
}

export async function check<T extends ECompilerType = ECompilerType.Rspack>(
	env: ITestEnv,
	context: ITestContext,
	name: string
) {
	const testConfig = context.getTestConfig();
	if (testConfig.noTests) return;

	const compiler = getCompiler(context, name);

	const errors: Array<{ message: string; stack?: string }> = (
		context.getError(name) || []
	).map(e => ({
		message: e.message,
		stack: e.stack
	}));
	const warnings: Array<{ message: string; stack?: string }> = [];

	const stats = compiler.getStats();
	const options = compiler.getOptions() as TCompilerOptions<T>;
	if (stats) {
		if (testConfig.writeStatsOuptut) {
			fs.writeFileSync(
				path.join(context.getDist(), "stats.txt"),
				stats.toString({
					preset: "verbose",
					colors: false
				}),
				"utf-8"
			);
		}

		if (testConfig.writeStatsJson) {
			const jsonStats = stats.toJson({
				errorDetails: true
			});
			fs.writeFileSync(
				path.join(context.getDist(), "stats.json"),
				JSON.stringify(jsonStats, null, 2),
				"utf-8"
			);
		}

		if (
			fs.existsSync(context.getSource("errors.js")) ||
			fs.existsSync(context.getSource("warnings.js")) ||
			stats.hasErrors() ||
			stats.hasWarnings()
		) {
			const statsJson = stats.toJson({
				errorDetails: true
			});
			if (statsJson.errors) {
				errors.push(...statsJson.errors);
			}
			if (statsJson.warnings) {
				warnings.push(...statsJson.warnings);
			}
		}
	}
	await checkArrayExpectation(
		context.getSource(),
		{ errors },
		"error",
		"errors",
		"Error",
		options
	);

	await checkArrayExpectation(
		context.getSource(),
		{ warnings },
		"warning",
		"warnings",
		"Warning",
		options
	);

	// clear error if checked
	if (fs.existsSync(context.getSource("errors.js"))) {
		context.clearError(name);
	}
}

export async function checkSnapshot<
	T extends ECompilerType = ECompilerType.Rspack
>(
	env: ITestEnv,
	context: ITestContext,
	name: string,
	snapshot: string,
	filter?: (file: string) => boolean
) {
	if (path.extname(snapshot) === ".snap") {
		throw new Error(
			"Snapshot with `.snap` will be managed by jest, please use `.snap.txt` instead"
		);
	}

	const compilerManager = getCompiler(context, name);
	const stats = compilerManager.getStats();
	const compiler = compilerManager.getCompiler();
	if (!stats || !compiler) return;

	const compilers: TCompiler<T>[] =
		"compilers" in compiler
			? (compiler.compilers as unknown as TCompiler<T>[])
			: [compiler as unknown as TCompiler<T>];
	const totalStats: TCompilerStats<T>[] =
		"stats" in stats
			? (stats.stats as unknown as TCompilerStats<T>[])
			: [stats as TCompilerStats<T>];
	const total = compilers.length;
	for (let i = 0; i < compilers.length; i++) {
		const c = compilers[i];
		const stats = totalStats[i];
		if (stats.hasErrors()) {
			const errors = [];
			errors.push(...stats.compilation.errors);

			throw new Error(
				`Failed to compile in fixture ${name}, Errors: ${errors
					?.map(i => `${i.message}\n${i.stack}`)
					.join("\n\n")}`
			);
		}
		const compilation =
			(c as unknown as TCompiler<ECompilerType.Rspack>)._lastCompilation ||
			(
				c as unknown as TCompiler<ECompilerType.Webpack> & {
					_lastCompilation: TCompilation<T>;
				}
			)._lastCompilation;

		const snapshotFileFilter =
			filter ||
			((file: string) =>
				(file.endsWith(".js") || file.endsWith(".mjs")) &&
				!file.includes("runtime.js"));

		const fileContents = Object.entries(compilation.assets)
			.filter(([file]) => snapshotFileFilter(file))
			.map(([file, source]) => {
				const tag = path.extname(file).slice(1) || "txt";
				let content = normalizePlaceholder(source.source().toString());
				const testConfig = context.getTestConfig();
				if (testConfig.snapshotContent) {
					content = testConfig.snapshotContent(content);
				}

				return `\`\`\`${tag} title=${file}\n${content}\n\`\`\``;
			});
		fileContents.sort();
		const content = fileContents.join("\n\n");
		const snapshotPath = path.isAbsolute(snapshot)
			? snapshot
			: path.resolve(
					context.getSource(),
					`./__snapshots__/${snapshot}${total > 1 ? `-${i}` : ""}`
				);

		env.expect(content).toMatchFileSnapshot(snapshotPath);
	}
}

export async function afterExecute(context: ITestContext, name: string) {
	const compiler = getCompiler(context, name);
	const testConfig = context.getTestConfig();
	if (typeof testConfig.afterExecute === "function") {
		let options = compiler.getOptions();
		if (Array.isArray(options) && options.length === 1) {
			options = options[0];
		}
		testConfig.afterExecute(options);
	}
}

export function findMultiCompilerBundle<
	T extends ECompilerType = ECompilerType.Rspack
>(
	context: ITestContext,
	name: string,
	multiFindBundle: (
		index: number,
		context: ITestContext,
		options: TCompilerOptions<T>
	) => string[] | string | void
) {
	if (typeof multiFindBundle !== "function") {
		return [];
	}

	const multiCompilerOptions = (context.getValue(
		name,
		"multiCompilerOptions"
	) || []) as TCompilerOptions<T>[];
	const result: string[] = [];
	const multiFileIndexMap: Record<string, number[]> =
		context.getValue(name, "multiFileIndexMap") || {};
	for (const [index, compilerOptions] of multiCompilerOptions.entries()) {
		const curBundles = multiFindBundle!(index, context, compilerOptions);

		const bundles = Array.isArray(curBundles)
			? curBundles
			: curBundles
				? [curBundles]
				: [];

		for (const bundle of bundles) {
			if (multiFileIndexMap[bundle]) {
				multiFileIndexMap[bundle].push(index);
			} else {
				multiFileIndexMap[bundle] = [index];
			}
		}

		result.push(...bundles);
	}

	context.setValue(name, "multiFileIndexMap", multiFileIndexMap);

	return result;
}

export function configMultiCompiler<
	T extends ECompilerType = ECompilerType.Rspack
>(
	context: ITestContext,
	name: string,
	configFiles: string[],
	defaultOptions: (index: number, context: ITestContext) => TCompilerOptions<T>,
	overrideOptions: (
		index: number,
		context: ITestContext,
		options: TCompilerOptions<T>
	) => void
) {
	const multiCompilerOptions: TCompilerOptions<T>[] = [];
	const caseOptions: TCompilerOptions<T>[] = Array.isArray(configFiles)
		? readConfigFile(
				configFiles!.map(i => context.getSource(i)),
				context,
				{},
				configs => {
					return configs.flatMap(c => {
						if (typeof c === "function") {
							const options = {
								testPath: context.getDist(),
								env: undefined
							};

							return c(options.env, options) as TCompilerOptions<T>;
						}

						return c as TCompilerOptions<T>;
					});
				}
			)
		: [{}];

	for (const [index, options] of caseOptions.entries()) {
		const compilerOptions = merge(
			typeof defaultOptions === "function"
				? defaultOptions!(index, context)
				: {},
			options
		);

		if (typeof overrideOptions === "function") {
			overrideOptions!(index, context, compilerOptions);
		}

		multiCompilerOptions.push(compilerOptions);
	}

	const compiler = getCompiler(context, name);
	compiler.setOptions(multiCompilerOptions as any);
	context.setValue(name, "multiCompilerOptions", multiCompilerOptions);
}
