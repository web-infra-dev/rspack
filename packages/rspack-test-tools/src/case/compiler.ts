import { ISimpleProcessorOptions, SimpleTaskProcessor } from "../processor";
import { getSimpleProcessorRunner } from "../test/simple";
import { ECompilerType } from "../type";

export type TCompilerCaseConfig = Omit<
	ISimpleProcessorOptions,
	"name" | "compilerType"
> & {
	description: string;
};

export function createCompilerCase(
	name: string,
	src: string,
	dist: string,
	testConfig: string
) {
	const caseConfig: TCompilerCaseConfig = require(testConfig);
	const runner = getSimpleProcessorRunner(src, dist);

	it(caseConfig.description, async () => {
		await runner(
			name,
			new SimpleTaskProcessor({
				name: name,
				compilerType: ECompilerType.Rspack,
				...caseConfig
			})
		);
	});
}
