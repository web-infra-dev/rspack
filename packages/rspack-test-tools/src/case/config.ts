import { RspackConfigProcessor } from "../processor/config";
import { MultipleRunnerFactory } from "../runner";
import { BasicCaseCreator } from "../test/creator";

const creator = new BasicCaseCreator({
	clean: true,
	describe: true,
	steps: ({ name }) => [
		new RspackConfigProcessor({
			name,
			runable: true
		})
	],
	runner: MultipleRunnerFactory
});

export function createConfigCase(name: string, src: string, dist: string) {
	creator.create(name, src, dist);
}
