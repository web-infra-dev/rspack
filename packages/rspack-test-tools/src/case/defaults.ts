import path from "path";

import {
	DefaultsConfigTaskProcessor,
	IDefaultsConfigProcessorOptions
} from "../processor";
import { TestContext } from "../test/context";
import { ITestEnv, ITestProcessor } from "../type";

export type TDefaultsCaseConfig = Omit<
	IDefaultsConfigProcessorOptions,
	"name"
> & {
	description: string;
};

const srcDir = path.resolve(__dirname, "../../tests/fixtures");
const distDir = path.resolve(__dirname, "../../tests/js/defaults");

const context = new TestContext({
	src: srcDir,
	dist: distDir
});

async function run(name: string, processor: ITestProcessor) {
	try {
		await processor.before?.(context);
		await processor.config?.(context);
	} catch (e: unknown) {
		context.emitError(name, e as Error);
	} finally {
		await processor.check?.(null as unknown as ITestEnv, context);
		await processor.after?.(context);
	}
}

export function createDefaultsCase(name: string, src: string) {
	const caseConfig = require(src) as TDefaultsCaseConfig;
	it(`should generate the correct defaults from ${caseConfig.description}`, async () => {
		await run(
			name,
			new DefaultsConfigTaskProcessor({
				name,
				...caseConfig
			})
		);
	});
}
