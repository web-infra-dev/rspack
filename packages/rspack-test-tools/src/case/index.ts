export * from "./config";
export * from "./diagnostic";
export * from "./diff";
export * from "./hash";
export * from "./hot";
export * from "./normal";
export * from "./stats";
export * from "./watch";
