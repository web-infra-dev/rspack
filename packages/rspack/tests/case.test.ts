import { describeCases } from "./case.template";

describeCases({
	name: "basic"
});
