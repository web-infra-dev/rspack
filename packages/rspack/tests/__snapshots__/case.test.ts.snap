// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cases exported tests should rewrite the css url() 1`] = `"b3523cb75fe70add.png"`;

exports[`cases exported tests should rewrite the css url() 2`] = `""data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAwAAAAOCAYAAAAbvf3sAAAACXBIWXMAABYlAAAWJQFJUiTwAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAEWSURBVHgBjZFRSsNAEIb/3fQAOUK9gfpe6YIovjU3UE+gJ2g8gTcw3sC+itUFrW+BeAN7ATFSKAjbHWcSC+mSlszLDrvfzvzzj0LHmNpZBtBQd4Gf7fuYgHPPueoCe1DK1UsHd7Dzw+P0daQj/SC5h09OzdGktw221vZXUBlVMN0ILPd6O9yzBBXX8CBdv6kWOGa4YLivQJNjM0ia73oTLsBwJjCB5gu4i7CgbsIey5ThEcOfKziTGFOGHypJeZ7jZ/Gbst4x2/fN9h2eGTNvgk92VrDEWLNmfJXLpIYrRy5b4Fs+9v2/pL0oUncI7FuHLI6PK5lJZGoe8qXNvrry27DesoRLpLPmNkTw9yEcxPWJMR+S/AFbfpAZqxwUNQAAAABJRU5ErkJggg==""`;

exports[`cases exported tests should rewrite the css url() with publicPath and ~@ prefix 1`] = `"/image/logo.b3523cb7.png"`;

exports[`cases exported tests should rewrite the css url() with publicPath when output.cssFilename is set 1`] = `"/image/logo.b3523cb7.png"`;
